onair/data_handling/sbn_adapter.py

```diff
@@ -0,0 +1,198 @@
+                struct_name = data_struct.__name__
+                # Skip the header, walk through the stuct
+                for field_name, field_type in data_struct._fields_[1:]:
+                    field_names = self.gather_field_names(app_name + "." + field_name, field_type)
+                    for field_name in field_names:
+                        self.currentData[x]['headers'].append(field_name)
+                        self.currentData[x]['data'].append([0]) #initialize all the data arrays with zero
+
+        return extract_meta_data_handle_ss_breakdown(meta_data_file, ss_breakdown)
+
+    def process_data_file(self, data_file):
+        print("SBN Adapter ignoring data file (telemetry should be live)")
+
+    def get_vehicle_metadata(self):
+        return self.all_headers, self.binning_configs['test_assignments']
+
+    def get_next(self):
+        """Provides the latest data from SBN in a dictionary of lists structure.
+        Returned data is safe to use until the next get_next call.
+        Blocks until new data is available."""
+
+        data_available = False
+
+        while not data_available:
+            with self.new_data_lock:
+                data_available = self.has_data()
+
+            if not data_available:
+                time.sleep(0.01)
+
+        read_index = 0
+        with self.new_data_lock:
+            self.new_data = False
+            self.double_buffer_read_index = (self.double_buffer_read_index + 1) % 2
+            read_index = self.double_buffer_read_index
+
+        return self.currentData[read_index]['data']
+
+    def has_more(self):
+        """Returns true if the adapter has more data.
+           For now always true: connection should be live as long as cFS is running.
+           TODO: allow to detect if cFS/the connection has died"""
+        return True
+
+    def message_listener_thread(self):
+        """Thread to listen for incoming messages from SBN"""
+
+        while(True):
+            generic_recv_msg_p = POINTER(sbn.sbn_data_generic_t)()
+            sbn.recv_msg(generic_recv_msg_p)
+
+            msgID = generic_recv_msg_p.contents.TlmHeader.Primary.StreamId
+            app_name, data_struct = self.msgID_lookup_table[msgID]
+
+            recv_msg_p = POINTER(data_struct)()
+            recv_msg_p.contents = generic_recv_msg_p.contents
+            recv_msg = recv_msg_p.contents
+
+            # prints out the data from the message to the terminal
+            print(", ".join([field_name + ": " + str(getattr(recv_msg, field_name)) for field_name, field_type in recv_msg._fields_[1:]]))
+
+            # TODO: Lock needed here?
+            self.get_current_data(recv_msg, data_struct, app_name)
+
+    def get_current_data(self, recv_msg, data_struct, app_name):
+        # TODO: Lock needed here?
+        current_buffer = self.currentData[(self.double_buffer_read_index + 1) %2]
+        secondary_header = recv_msg.TlmHeader.Secondary
```

Comment:
Why did you add back in the timing field?

Reply:
Oh, it looks like I removed that in a different version. Apologies!

We don't want to force any assumptions about time, so these lines should all be removed:

```
         secondary_header = recv_msg.TlmHeader.Secondary
         #gets seconds from header and adds to current buffer
         start_time = datetime.datetime(1969, 12, 31, 20) # January 1, 1980
         seconds = secondary_header.Seconds
         subseconds = secondary_header.Subseconds
         curr_time = seconds + (2**(-32) * subseconds) # a subsecond is equal to 2^-32 second
         time = start_time + datetime.timedelta(seconds=curr_time)
         str_time = time.strftime("%Y-%j-%H:%M:%S.%f")
         current_buffer['data'][0] = str_time
```

Reply:
Thanks for removing!